zb-db/src/main/java/io/zeebe/db/impl/rocksdb/transaction/ZeebeTransactionDb.java

```diff
@@ -200,17 +218,13 @@ private boolean isRocksDbExceptionRecoverable(RocksDBException rdbex) {
   }
 
   private void runInNewTransaction(TransactionOperation operations) throws Exception {
-    try (WriteOptions options = new WriteOptions()) {
-      currentTransaction = new ZeebeTransaction(optimisticTransactionDB.beginTransaction(options));
-
+    try {
+      inTransaction = true;
       operations.run();
-
       currentTransaction.commit();
     } finally {
-      if (currentTransaction != null) {
-        currentTransaction.close();
-        currentTransaction = null;
-      }
+      currentTransaction.rollback();
```

Comment:
What happens if rollback throws an exception, shouldn't we handle this? Is the transaction then still reusable?

Reply:
This is currently catched in the `#transaction` method. I will change this behavior and retry the commit or rollback on error with the next PR related to this #2201 .

Reply:
If it is catched outside this method, how is then `inTransaction` reset, as the only way to reach `runInNewTransaction` is if `inTransaction` is `false`. Also my main question is can a transaction end in a state where it is not reusable anymore? Maybe if we fail to rollback, we should discard the transaction and create a new one?

Reply:
Had a look at the `c++` code and they simply clear everything and return OK

```c++
Status OptimisticTransaction::Rollback() {
  Clear();
  return Status::OK();
}
```

So there is no way that an RocksDBException can be thrown, since on the JNI layer they only throw exceptions if the StatusCode is different then OK.